# Unit 12 - Tales from the Crypto

---


## 1. Sentiment Analysis

Use the [newsapi](https://newsapi.org/) to pull the latest news articles for Bitcoin and Ethereum and create a DataFrame of sentiment scores for each coin.

Use descriptive statistics to answer the following questions:
1. Which coin had the highest mean positive score?
2. Which coin had the highest negative score?
3. Which coin had the highest positive score?

In [110]:
# Initial imports
import os
import pandas as pd
from dotenv import load_dotenv
import nltk as nltk
nltk.download('vader_lexicon')
from nltk.sentiment.vader import SentimentIntensityAnalyzer
analyzer = SentimentIntensityAnalyzer()

%matplotlib inline

[nltk_data] Downloading package vader_lexicon to
[nltk_data]     C:\Users\ferna_nlq0xu3\AppData\Roaming\nltk_data...
[nltk_data]   Package vader_lexicon is already up-to-date!


In [119]:
# Read your api key environment variable
# YOUR CODE HERE!
from newsapi import NewsApiClient
load_dotenv()
api_key = os.getenv('NEWS_API')

In [120]:
# Create a newsapi client
# YOUR CODE HERE!
newsapi = NewsApiClient(api_key=api_key)

In [145]:
# Fetch the Bitcoin news articles
# YOUR CODE HERE!
btc_articles = newsapi.get_everything(q="bitcoin", 
                                      language= "en", 
                                      sort_by= "relevancy",
                                      page_size=100)

btc_articles['totalResults']



7610

In [146]:
# Fetch the Ethereum news articles
# YOUR CODE HERE!
eth_articles = newsapi.get_everything(q='ethereum',
                                      language='en',
                                      sort_by='relevancy',
                                     page_size=100)
eth_articles['totalResults']


2483

In [153]:
# Create the Bitcoin sentiment scores DataFrame
# YOUR CODE HERE!
btc_sentiment = []
for article in btc_articles["articles"]:
    text = article["content"]
    sentiment = analyzer.polarity_scores(text)
    compound = sentiment["compound"]
    positive = sentiment["pos"]
    neutral = sentiment["neu"]
    negative = sentiment["neg"]
    dictionary = {"Compound": compound, "Negative": negative, "Neutral": neutral, "Positive": positive, "text": text}
    btc_sentiment.append(dictionary)     
btc_df = pd.DataFrame(btc_sentiment)
btc_df


    

Compound  Negative  Neutral  Positive  \
0     0.6908     0.000    0.822     0.178   
1    -0.5719     0.184    0.705     0.111   
2    -0.6124     0.143    0.857     0.000   
3     0.6240     0.000    0.873     0.127   
4     0.7264     0.000    0.836     0.164   
..       ...       ...      ...       ...   
95   -0.5423     0.123    0.877     0.000   
96    0.0000     0.000    1.000     0.000   
97    0.3818     0.000    0.927     0.073   
98   -0.3400     0.070    0.930     0.000   
99    0.1027     0.076    0.812     0.113   

                                                 text  
0   Like Dogecoin devotees, the mayor of Reno, and...  
1   Filed under:\r\nThe supply chain attack has re...  
2   image copyrightGetty Images\r\nThe gang behind...  
3   To get a roundup of TechCrunchs biggest and mo...  
4   While retail investors grew more comfortable b...  
..                                                ...  
95  Metropolitan police detectives investigating i...  
96  A representation of the virtual cryptocurrency...  
97  Speaking remotely at the B Word Conference tod...  
98  This article was translated from our Spanish e...  
99  Amazon on Monday denied a media report that th...  

[100 rows x 5 columns]

In [154]:
# Create the Ethereum sentiment scores DataFrame
# YOUR CODE HERE!
eth_sentiment = []
for article in eth_articles["articles"]:
    text = article["content"]
    sentiment = analyzer.polarity_scores(text)
    compound = sentiment["compound"]
    positive = sentiment["pos"]
    neutral = sentiment["neu"]
    negative = sentiment["neg"]
    dictionary = {"Compound": compound, "Negative": negative, "Neutral": neutral, "Positive": positive, "text": text}
    eth_sentiment.append(dictionary)     
eth_df = pd.DataFrame(eth_sentiment)
eth_df

Compound  Negative  Neutral  Positive  \
0     0.3612     0.000    0.925     0.075   
1    -0.2411     0.061    0.939     0.000   
2     0.7264     0.000    0.836     0.164   
3     0.6956     0.000    0.810     0.190   
4     0.0000     0.000    1.000     0.000   
..       ...       ...      ...       ...   
95    0.0000     0.000    1.000     0.000   
96    0.3818     0.000    0.893     0.107   
97    0.5859     0.000    0.879     0.121   
98    0.1027     0.057    0.876     0.067   
99    0.0000     0.000    1.000     0.000   

                                                 text  
0   There are many blockchain platforms competing ...  
1   Blockchain infrastructure startups are heating...  
2   While retail investors grew more comfortable b...  
3   Cent was founded in 2017 as an ad-free creator...  
4   Representation of the Ethereum virtual currenc...  
..                                                ...  
95  Representations of cryptocurrencies Bitcoin, E...  
96  LONDON/HONG KONG, Aug 5 (Reuters) - World stoc...  
97  Bitcoin extended gains past $32,000, continuin...  
98  image copyrightGetty Images\r\nBitcoin has jum...  
99  Two versions of a rare job application filled ...  

[100 rows x 5 columns]

In [156]:
# Describe the Bitcoin Sentiment
# YOUR CODE HERE!
btc_df.describe()

Compound    Negative     Neutral    Positive
count  100.000000  100.000000  100.000000  100.000000
mean     0.091856    0.033960    0.906890    0.059140
std      0.394892    0.058909    0.087044    0.070527
min     -0.827100    0.000000    0.653000    0.000000
25%     -0.044250    0.000000    0.852500    0.000000
50%      0.000000    0.000000    0.919500    0.047000
75%      0.381800    0.066250    1.000000    0.092000
max      0.883400    0.287000    1.000000    0.276000

In [157]:
# Describe the Ethereum Sentiment
# YOUR CODE HERE!
eth_df.describe()

Compound   Negative     Neutral    Positive
count  100.000000  100.00000  100.000000  100.000000
mean     0.154366    0.02012    0.923940    0.055920
std      0.330502    0.04262    0.073756    0.063166
min     -0.812600    0.00000    0.714000    0.000000
25%      0.000000    0.00000    0.872500    0.000000
50%      0.000000    0.00000    0.929000    0.045000
75%      0.401900    0.00000    1.000000    0.094500
max      0.771700    0.24900    1.000000    0.201000

### Questions:

Q: Which coin had the highest mean positive score?

A: Bitcoin : 0.59140

Q: Which coin had the highest compound score?

A: Bitcoin : 0.883400

Q. Which coin had the highest positive score?

A: Bitcoin : 0.276000

---

## 2. Natural Language Processing
---
###   Tokenizer

In this section, you will use NLTK and Python to tokenize the text for each coin. Be sure to:
1. Lowercase each word.
2. Remove Punctuation.
3. Remove Stopwords.

In [160]:
from nltk.tokenize import word_tokenize, sent_tokenize
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer, PorterStemmer
from string import punctuation
import re

In [161]:
# Instantiate the lemmatizer
# YOUR CODE HERE!
lemmatizer = WordNetLemmatizer()

# Create a list of stopwords
# YOUR CODE HERE!

# Expand the default stopwords list if necessary
# YOUR CODE HERE!




In [165]:
# Complete the tokenizer function
def tokenizer(text):
    regex = re.compile("[^a-zA-Z ]")
    re_clean = regex.sub('', text)

    # Remove the punctuation from text
    words = [t.translate(str.maketrans('','',string.punctuation)) for t in tokens]
   
    # Create a tokenized list of the words
    tokens = word_tokenize(re_clean)
    
    # Lemmatize words into root words
    tokens = ', '.join([lemmatizer.lemmatize(word) for word in words])
    return tokens

    # Convert the words to lowercase
    tokens = [t.lower() for t in tokens]
    
    
    # Remove the stop words
    stop_words = set(stopwords.words('english'))
    words = [w for w in words if not w in stop_words]


In [182]:

# Compile the list of lists of tokens for each Bitcoin article
# YOUR CODE HERE!
btc_df["tokens"]= btc_word_tokenized
btc_df.head



ValueError: Length of values (0) does not match length of index (100)

In [14]:
# Create a new tokens column for Ethereum
# YOUR CODE HERE!

---

### NGrams and Frequency Analysis

In this section you will look at the ngrams and word frequency for each coin. 

1. Use NLTK to produce the n-grams for N = 2. 
2. List the top 10 words for each coin. 

In [175]:
from collections import Counter
from nltk import ngrams

In [185]:
# Generate the Bitcoin N-grams where N=2
# YOUR CODE HERE!
corpus = btc_articles["description"]
bigram_counter(corpus)


KeyError: 'description'

In [186]:
# Generate the Ethereum N-grams where N=2
# YOUR CODE HERE!

In [18]:
# Function token_count generates the top 10 words for a given coin
def token_count(tokens, N=3):
    """Returns the top N tokens from the frequency count"""
    return Counter(tokens).most_common(N)

In [19]:
# Use token_count to get the top 10 words for Bitcoin
# YOUR CODE HERE!

In [20]:
# Use token_count to get the top 10 words for Ethereum
# YOUR CODE HERE!

---

### Word Clouds

In this section, you will generate word clouds for each coin to summarize the news for each coin

In [21]:
from wordcloud import WordCloud
import matplotlib.pyplot as plt
plt.style.use('seaborn-whitegrid')
import matplotlib as mpl
mpl.rcParams['figure.figsize'] = [20.0, 10.0]

In [22]:
# Generate the Bitcoin word cloud
# YOUR CODE HERE!

In [23]:
# Generate the Ethereum word cloud
# YOUR CODE HERE!

---
## 3. Named Entity Recognition

In this section, you will build a named entity recognition model for both Bitcoin and Ethereum, then visualize the tags using SpaCy.

In [24]:
import spacy
from spacy import displacy

In [25]:
# Download the language model for SpaCy
# !python -m spacy download en_core_web_sm

In [26]:
# Load the spaCy model
nlp = spacy.load('en_core_web_sm')

---
### Bitcoin NER

In [27]:
# Concatenate all of the Bitcoin text together
# YOUR CODE HERE!

In [28]:
# Run the NER processor on all of the text
# YOUR CODE HERE!

# Add a title to the document
# YOUR CODE HERE!

In [29]:
# Render the visualization
# YOUR CODE HERE!

In [30]:
# List all Entities
# YOUR CODE HERE!

---

### Ethereum NER

In [31]:
# Concatenate all of the Ethereum text together
# YOUR CODE HERE!

In [32]:
# Run the NER processor on all of the text
# YOUR CODE HERE!

# Add a title to the document
# YOUR CODE HERE!

In [33]:
# Render the visualization
# YOUR CODE HERE!

In [34]:
# List all Entities
# YOUR CODE HERE!

---